File: apps/frontend/src/app/foundation/reservations/reservation.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { ApiClient, ApiPage } from '@foundation/networking';
import {
  BookingDetail,
  CreateReservationRequest,
  Reservation,
  ReservationQuote,
  ReservationQuoteRequest,
} from './reservation.models';

@Injectable({ providedIn: 'root' })
export class ReservationService {
  private readonly api = inject(ApiClient);

  listMine(page = 0, size = 20): Observable<ApiPage<Reservation>> {
    return this.api.get<ApiPage<Reservation>>('/reservations/me', {
      params: {
        page,
        size,
      },
    });
  }

  getReservation(reservationId: string): Observable<Reservation> {
    return this.api.get<Reservation>(`/reservations/${encodeURIComponent(reservationId)}`);
  }

  getReservationDetail(reservationId: string): Observable<BookingDetail> {
    return this.api.get<BookingDetail>(`/reservations/${encodeURIComponent(reservationId)}/detail`);
  }

  quote(payload: ReservationQuoteRequest): Observable<ReservationQuote> {
    return this.api.post<ReservationQuote, ReservationQuoteRequest>('/reservations/quote', payload);
  }

  create(payload: CreateReservationRequest): Observable<Reservation> {
    return this.api.post<Reservation, CreateReservationRequest>('/reservations', payload);
  }

  cancel(reservationId: string, reason?: string): Observable<Reservation> {
    return this.api.post<Reservation, { reason?: string }>(
      `/reservations/${encodeURIComponent(reservationId)}/cancel`,
      { reason },
    );
  }
}

File: apps/ios/SpotLink/Sources/SpotLink/Features/Reservations/ReservationService.swift
import Foundation

// MARK: - Reservation Service

public final class ReservationService: Sendable {
    private let apiClient: APIClientProtocol

    public init(apiClient: APIClientProtocol) {
        self.apiClient = apiClient
    }

    public func listMine(page: Int = 0, size: Int = 20) async throws -> APIPage<Reservation> {
        try await apiClient.get("/reservations/me", query: ["page": String(page), "size": String(size)])
    }

    public func getReservation(_ id: String) async throws -> Reservation {
        try await apiClient.get("/reservations/\(id)", query: nil)
    }

    public func getReservationDetail(_ id: String) async throws -> BookingDetail {
        try await apiClient.get("/reservations/\(id)/detail", query: nil)
    }

    public func quote(_ request: ReservationQuoteRequest) async throws -> ReservationQuote {
        try await apiClient.post("/reservations/quote", body: request)
    }

    public func create(_ request: CreateReservationRequest) async throws -> Reservation {
        try await apiClient.post("/reservations", body: request)
    }

    public func cancel(_ id: String, reason: String? = nil) async throws -> Reservation {
        try await apiClient.post("/reservations/\(id)/cancel", body: CancelReservationRequest(reason: reason))
    }
}
